llvm/lib/CASObjectFormats/FlatV1.cpp

```diff
@@ -436,28 +424,43 @@ Expected<BlockRef> BlockRef::create(CompileUnitBuilder &CUB,
     Fixups.back().Offset = E->getOffset();
   }
 
-  // Normalize content and put it at the end of the block.
-  // FIXME: assume current block alignment is the alignment for padding and just
-  // pad every block.
-  Optional<Align> TrailingNopsAlignment;
-  TrailingNopsAlignment = Align{Block.getAlignment()};
-  StringRef Content(Block.getContent().begin(), Block.getSize());
-  SmallString<1024> MutableContentStorage;
-  if (Error E = helpers::canonicalizeContent(CUB.TT.getArch(), Content,
-                                             MutableContentStorage, Fixups,
-                                             TrailingNopsAlignment)
-                    .moveInto(Content))
-    return E;
-
+  // Create BlockData.
+  SmallString<1024> EncodeContent;
+  if (!Block.isZeroFill()) {
+    SmallString<1024> MutableContentStorage;
+    // Normalize content and put it at the end of the block.
+    // FIXME: assume current block alignment is the alignment for padding and
+    // just pad every block.
+    Optional<Align> TrailingNopsAlignment;
+    TrailingNopsAlignment = Align{Block.getAlignment()};
+    StringRef Content(Block.getContent().begin(), Block.getSize());
+    if (Error E = helpers::canonicalizeContent(CUB.TT.getArch(), Content,
+                                               MutableContentStorage, Fixups,
+                                               TrailingNopsAlignment)
+                      .moveInto(Content))
+      return E;
+    data::BlockData::encode(Content.size(), Block.getAlignment(),
+                            Block.getAlignmentOffset(), Content,
+                            ArrayRef<Fixup>(), EncodeContent);
+  } else
+    data::BlockData::encode(Block.getSize(), Block.getAlignment(),
+                            Block.getAlignmentOffset(), None,
+                            ArrayRef<Fixup>(), EncodeContent);
```

Comment:
I'd probably structure this to avoid the duplicate call to `BlockData::encode`:
```
Optional<StringRef> Content;
size_t BlockSize = Block.size();
if (!Block.isZeroFill()) {
  if (Error E = canonicalize().moveInto(Content))
   return E;
  BlockSize = Content->size();
}
data::BlockData::encode(BlockSize, ..., Content, Fixups);
```